pkg/backend/snapshot.go

```diff
@@ -662,23 +664,55 @@ func (sm *SnapshotManager) snap() *deploy.Snapshot {
 		secretsManager = sm.baseSnapshot.SecretsManager
 	}
 
+	var metadata deploy.SnapshotMetadata
+	if sm.baseSnapshot != nil {
+		metadata = sm.baseSnapshot.Metadata
+	}
+
 	manifest.Magic = manifest.NewMagic()
-	return deploy.NewSnapshot(manifest, secretsManager, resources, operations)
+	return deploy.NewSnapshot(manifest, secretsManager, resources, operations, metadata)
 }
 
-// saveSnapshot persists the current snapshot and optionally verifies it afterwards.
+// saveSnapshot persists the current snapshot. If integrity checking is enabled,
+// the snapshot's integrity is also verified. If the snapshot is invalid,
+// metadata about this write operation is added to the snapshot before it is
+// written, in order to aid debugging should future operations fail with an
+// error.
 func (sm *SnapshotManager) saveSnapshot() error {
 	snap, err := sm.snap().NormalizeURNReferences()
 	if err != nil {
 		return fmt.Errorf("failed to normalize URN references: %w", err)
 	}
+
+	// In order to persist metadata about snapshot integrity issues, we check the
+	// snapshot's validity *before* we write it. However, should an error occur,
+	// we will only raise this *after* the write has completed. In the event that
+	// integrity checking is disabled, we still actually perform the check (and
+	// write metadata appropriately), but we will not raise the error following a
+	// successful write.
+	//
+	// If the actual write fails for any reason, this error will supersede any
+	// integrity error. This matches behaviour prior to when integrity metadata
+	// writing was introduced.
+	//
+	// Metadata will be cleared out by a successful operation (even if integrity
+	// checking is being enforced).
+	integrityError := snap.VerifyIntegrity()
+	if integrityError == nil {
+		snap.Metadata.IntegrityErrorMetadata = nil
+	} else {
+		snap.Metadata.IntegrityErrorMetadata = &deploy.SnapshotIntegrityErrorMetadata{
```

Comment:
I was wondering if we should keep a list of recent errors here.  However I *think* the user would get a snapshot integrity error shown before they would even be able to try and save the snapshot, and never get here even for write operations, so this looks good.